src/MongoDB.Analyzer/Core/Builders/BuildersResolveVariables.cs

```diff
@@ -0,0 +1,492 @@
+            var generateNewContext = isCompoundExpression || expressionAnalysisContext == null;
+
+            if (generateNewContext)
+            {
+                expressionAnalysisContext = GenerateNewExpressionAnalysisContext(assignmentExpression, syntaxNodeToVariableValueMap, contextInformation);
+            }
+
+            var variableValue = generateNewContext ? new VariableValue(expressionAnalysisContext, new List<Location>()) : rightVariableValue;
+            dirtyVariables.Add(variableName);
+            variableValues[variableName] = variableValue;
+            var originalExpression = variableValue.ExpressionAnalysisContext.Node.OriginalExpression;
+
+            if (builderNodesProcessed.Add(originalExpression))
+            {
+                variableValuesList.Add(variableValue);
+            }
+
+            if (assignmentExpression.Parent is ConditionalExpressionSyntax)
+            {
+                RemoveVariablesFromContext(new List<string>() { variableName}, contextInformation);
+            }
+        }
+        else
+        {
+            RemoveVariablesFromContext(new List<string>() { variableName }, contextInformation);
+        }
+
+        return canEvaluate;
+    }
+
+    private static void RemoveVariablesFromContext(IEnumerable<string> variablesToRemove, ContextInformation contextInformation)
+    {
+        var variableValues = contextInformation.VariableValues;
+        var variableValuesList = contextInformation.VariableValuesList;
+        var builderNodesProcessed = contextInformation.BuilderNodesProcessed;
+
+        foreach (var variable in variablesToRemove)
+        {
+            var isValidVariable = variableValues.TryGetValueSafe(variable, out var previousValue);
+
+            if (!isValidVariable)
+            {
+                continue;
+            }
+
+            var originalExpression = previousValue.ExpressionAnalysisContext.Node.OriginalExpression;
+
+            if (builderNodesProcessed.Add(originalExpression))
+            {
+                variableValuesList.Add(previousValue);
+            }
+        }
+
+        variableValues.RemoveRange(variablesToRemove);
+    }
+
+    private static bool TryProcessBuildersExpression(SyntaxNode syntaxNode, ContextInformation contextInformation, HashSet<string> dirtyVariables)
+    {
+        var isValidBuildersExpression = false;
+        var semanticModel = contextInformation.SemanticModel;
+
+        if (syntaxNode is VariableDeclaratorSyntax variableDeclarator)
+        {
+            var leftHandSide = variableDeclarator.Identifier;
+            var rightHandSide = variableDeclarator.Initializer.Value;
+            var argumentList = variableDeclarator.ArgumentList;
+            var variableDeclaration = variableDeclarator.Parent as VariableDeclarationSyntax;
+            var symbol = semanticModel.GetSymbolInfo(variableDeclaration.Type).Symbol as ITypeSymbol;
```

Comment:
Make sure only the vars needed in more than one of the code paths, are defined.
`symbol `is used?

Reply:
Done